### PR TITLE
[feat] engine: implementation of cppreference

### PR DESCRIPTION
**Pull request by thunder-coding**
## What does this PR do?
Implement cppreference engine

## Why is this change important?
Search for C++ reference documentation

## How to test this PR locally?
Start dev server and search for `std::vector`, `std::string` or any other C++ reference

## Author's checklist
- [X] search functionality works

Also cppreference does support a bunch of other languages, but they aren't complete. Pages are missing, as well as content may be incomplete. I don't know any of those languages in there. Also I don't have any experience working with localization and stuff, would be nice if anyone else does take care of that.

## Related issues
None

### CODE REVIEW

**Comment by return42** [comment]
> would be nice if anyone else does take care of that.

`cppreference.com` is a [MediaWiki](https://www.mediawiki.org/wiki/MediaWiki) / we have a (abstract) [MediaWiki engine](https://docs.searxng.org/dev/engines/mediawiki.html) that supports languages:

https://github.com/searxng/searxng/blob/c0b97c65438be44dd5f5e64786576763993ce424/searx/engines/mediawiki.py#L97-L98

May you want to give it a try / its very easy to setup, here is an example from the `settings.yml` file:

https://github.com/searxng/searxng/blob/c0b97c65438be44dd5f5e64786576763993ce424/searx/settings.yml#L735-L745

If you look out for `engine: mediawiki` you will find a lot of other examples.

**Comment by thunder-coding** [comment]
@return42 the problem is that searxng's mediawiki engine assumes a lot of stuff. Like it appends `/w/` to the base url, cppreference uses `/mwiki/` instead of `/w/`, also addition of extra parameters seem to confuse cppreference even if I remove the `/w/` part

**Comment by unixfox** [comment]
Please disable the engine by default.

**Comment by return42** [comment]
> @return42 the problem is that searxng's mediawiki engine assumes a lot of stuff. Like it appends `/w/` to the base url, cppreference uses `/mwiki/` instead of `/w/`, also addition of extra parameters seem to confuse cppreference even if I remove the `/w/` part

TL;DR; API seems not suitable / Here is what I tried (without luck):

```diff
diff --git a/searx/engines/mediawiki.py b/searx/engines/mediawiki.py
index 6a9ac974a..cd7835a35 100644
--- a/searx/engines/mediawiki.py
+++ b/searx/engines/mediawiki.py
@@ -101,6 +101,8 @@ base_url: str = 'https://{language}.wikipedia.org/'
   ISO 639-1 language code (en, de, fr ..) of the search language.
 """
 
+api_path = 'w/api.php'
+
 timestamp_format = '%Y-%m-%dT%H:%M:%SZ'
 """The longhand version of MediaWiki time strings."""
 
@@ -115,9 +117,9 @@ def request(query, params):
         params['language'] = params['language'].split('-')[0]
 
     if base_url.endswith('/'):
-        api_url = base_url + 'w/api.php?'
+        api_url = base_url + f'{api_path}?'
     else:
-        api_url = base_url + '/w/api.php?'
+        api_url = base_url + f'/{api_path}?'
     api_url = api_url.format(language=params['language'])
 
     offset = (params['pageno'] - 1) * number_of_results
diff --git a/searx/settings.yml b/searx/settings.yml
index 72cf4e7ac..4136dc0f6 100644
--- a/searx/settings.yml
+++ b/searx/settings.yml
@@ -1895,6 +1895,16 @@ engines:
     categories: images
     number_of_results: 10
 
+  - name: cppreference
+    engine: mediawiki
+    shortcut: cpp
+    search_type: title
+    categories: [it]
+    base_url: "https://{language}.cppreference.com/"
+    api_path: "mwiki/api.php"
+    srprop: "snippet|timestamp"
+
+
   - name: wolframalpha
```

with I get URLs like https://de.cppreference.com/mwiki/api.php?action=query&list=search&format=json&srsearch=c_str&sroffset=0&srlimit=5&srwhat=title&srprop=snippet%7Ctimestamp&srsort=relevance&srenablerewrites=1

grafik

Not sure, but to me it looks like the mediawiki api of cppreference.com is old

- https://en.cppreference.com/mwiki/api.php?action=help ...

grafik

compared to the actual one:

- https://www.mediawiki.org/w/api.php?action=help

grafik
